module/preference/PreferenceDialog.kt

```diff
@@ -0,0 +1,386 @@
+                        placeholder = placeholder,
+                        border = border,
+                        onValueChange = { value = it; state.modified = true }
+                    )
+                }
+            }
+        }
+
+        class Checkbox(val state: PreferencesForm,
+                       initialValue: Boolean, override var label: String,
+                       override val caption: Optional<String> = Optional.empty()) : PreferenceField {
+
+            var value by mutableStateOf(initialValue)
+
+            @Composable
+            override fun Display() {
+                FieldLayout @Composable {
+                    Checkbox(
+                        value = value,
+                        onChange = { value = it; state.modified = true }
+                    )
+                }
+            }
+
+            override fun isValid(): Boolean {
+                return true
+            }
+        }
+
+        class Dropdown<T : Any>(val state: PreferencesForm, val values: List<T>, override val label: String,
+                                override val caption: Optional<String> = Optional.empty()) : PreferenceField {
+
+            var selected by mutableStateOf(values.first())
+
+            @Composable
+            override fun Display() {
+                FieldLayout @Composable {
+                    Form.Dropdown(
+                        values = values,
+                        selected = selected,
+                        onSelection = { selected = it; state.modified = true }
+                    )
+                }
+            }
+
+            override fun isValid(): Boolean {
+                return true
+            }
+        }
+    }
+
+    class PreferencesForm : State {
+        private val QUERY_LIMIT_PLACEHOLDER = "1000"
+        private val IGNORED_PATHS_PLACEHOLDER = ".git"
+        var modified by mutableStateOf(false)
+
+        // Graph Visualiser Preferences
+        var graphOutput = PreferenceField.Checkbox(
+            this, initialValue = preferenceMgr.graphOutputEnabled, label = ENABLE_GRAPH_OUTPUT,
+            caption = Optional.of(PREFERENCES_GRAPH_OUTPUT_CAPTION)
+        )
+
+        // Project Manager Preferences
+        private val ignoredPathsString = preferenceMgr.ignoredPaths.joinToString(",")
+        var ignoredPaths = PreferenceField.TextInput(
+            this, initialValue = ignoredPathsString,
+            label = PROJECT_IGNORED_PATHS, placeholder = IGNORED_PATHS_PLACEHOLDER
+        )
+
+        // Query Runner Preferences
+        var matchQueryLimit = PreferenceField.TextInput(
+            this, initialValue = preferenceMgr.matchQueryLimit.toString(),
+            label = SET_QUERY_LIMIT, placeholder = QUERY_LIMIT_PLACEHOLDER,
+            caption = Optional.of(PREFERENCES_MATCH_QUERY_LIMIT_CAPTION)
+        ) {/* validator = */ it.toLongOrNull() != null }
+        
+        // Text Editor Preferences
+        var autoSave = PreferenceField.Checkbox(
```

Comment:
So, I do note that currently each preference group _is_ pretty small (one element each?), but still, all this detail is just a tiny bit too much to stuff into a single class's fields. It becomes a little unwieldy.

Two possible solutions:
1. Extract each preference field initialiser into its own method.
2. Convert each `PreferenceGroup` to its own inner class: `PreferenceGroup.GraphVisualiser`, `PreferenceGroup.ProjectManager`...

I believe option 2 scales better.